docs(signals): add testing guide (#4461)
ngrx/platform: projects/ngrx.io/content/guide/signals/signal-store/testing.md

# Testing

There are two scenarios when it comes to testing:

1. Testing the `signalStore` itself,
2. Testing a component/service which uses the Signal Store.

Whereas in the first scenario, mocking will happen for dependencies of the Signal Store, scenario 2 might require to mock the Signal Store itself.

The examples are using the Jest API, but the same principles apply to other testing frameworks.

## Testing the Signal Store

### Testing without `TestBed`

The Signal Store is a function that returns a class. That means a test can instantiate the class and test it without the TestBed:

> 

Reviewer: I initially suggested providing both guides - with and without `TestBed`. After reconsideration, I think it would be enough to have only the guide with `TestBed` because:

- Without `TestBed` we cannot test the `onDestroy` hook.
- SignalStore can be tested without TestBed only when there are no injected services, reactive methods, and/or effects which is a rare case.

Author: Yeah, I will shorten it, but I think it is important to explain why our guide contains only the `TestBed` scenario (you mentioned some reasons already)

Reviewer: Yes, it can be mentioned. E.g:

```md
## Testing a SignalStore

Guide with `TestBed` ...

<div class="alert is-helpful">

SignalStore can also be tested without `TestBed`...

</div>
```

The `MovieComponent` utilizes the `MoviesStore` to display movies:

<code-example header="movies.component.ts">

@Component({
  selector: 'app-movies',
  template: ` <input type="text" [(ngModel)]="studio" [disabled]="store.loading()" placeholder="Name of Studio" />
    <ul>
      @for (movie of store.movies(); track movie.id) {
      <p>{{ movie.id }}: {{ movie.name }}</p>
      }
    </ul>`,
  standalone: true,
  imports: [FormsModule],
})
export class MoviesComponent {
  protected studio = signal('');
  protected readonly store = inject(MoviesStore);

  constructor() {
    this.store.load(this.studio);
  }
}

</code-example>

### Native Mocking

<code-example header="movies.component.spec.ts">

it('should show movies (native Jest)', () => {
  const load = jest.fn<void, [Signal<string>]>();

  const moviesStore = {
    movies: signal(new Array<Movie>()),
    loading: signal(false),
    load,
  };

  TestBed.configureTestingModule({
    imports: [MoviesComponent],
    providers: [
      {
        provide: MoviesStore,
        useValue: moviesStore,
      },
    ],
  });

  const fixture = TestBed.createComponent(MoviesComponent);
  fixture.autoDetectChanges(true);

>   const studio = load.mock.calls[0][0];

Reviewer 2: I think might be wrong.  Only one `[0]` ?

Author: `[0][0]` means the first parameter of the first call.

Reviewer 2: Yes, you are right. `mock.calls` returns an array of arrays.

        const startedAt = store._startedAt();
        if (!startedAt || store._status() === 'stopped') {
          return;
        }

        const timeSpent = new Date().getTime() - startedAt.getTime();
        const alreadySpent = store.trackedData()[store._currentId()] ?? 0;
        patchState(store, (state) => ({
          _currentId: 0,
          _status: 'stopped' as const,
          trackedData: { ...state.trackedData, [state._currentId]: alreadySpent + timeSpent },
        }));
      };

      return {
        play(id: number) {
          stop();
          patchState(store, {
            _currentId: id,
            _status: 'playing',
            _startedAt: new Date(),
          });
        },
        stop,
      };
    })
  );

</code-example>

There are two options for testing this extension: in combination with the `MoviesStore` or in isolation.

When tested with the `MoviesStore`, the same approach as in previous examples is followed.

To test the extension in isolation, an artificial "Wrapper" SignalStore is created. The test process remains straightforward.

<code-example header="with-play-tracking.spec.ts">

describe('withTrackedPlay', () => {

>   const TrackedPlayStore = signalStore({providedIn: 'root'}, withPlayTracking();

Reviewer 2: Missing closing parentheses.  And lines 662 - 670 need to add semi-colons.

> <!-- TOC -->

Reviewer: To make this page visible in the side nav, we should add link to the `navigation.json`: https://github.com/ngrx/platform/blob/main/projects/ngrx.io/content/navigation.json#L311

    * [What to test](#what-to-test)
    * [TestBed or not?](#testbed-or-not)
  * [Testing the Signal Store](#testing-the-signal-store)
    * [Globally provided](#globally-provided)
    * [Locally provided](#locally-provided)
    * [`withComputed`](#withcomputed)
    * [`withMethods`, DI, and asynchronous tasks](#withmethods-di-and-asynchronous-tasks)
    * [`rxMethod`](#rxmethod)
      * [with Observables](#with-observables)
      * [with Signals](#with-signals)
    * [Custom extensions](#custom-extensions)
  * [Mocking the Signal Store](#mocking-the-signal-store)
    * [Native Mocking](#native-mocking)
    * [ng-mocks](#ng-mocks)
    * [Partial Mocking via Spies](#partial-mocking-via-spies)
<!-- TOC -->

This is still a draft. Markdown is used for better readability. The rendered version of this guide is available at https://github.com/rainerhahnekamp/ngrx/blob/docs/signals/testing/projects/ngrx.io/content/guide/signals/signal-store/testing.md.

Once the content is finalized, we will replace markdown with `<code>` tags.

The examples used in this guide are available at https://github.com/rainerhahnekamp/ngrx-signal-store-testing

## Introduction

> ### On Testing in General

Reviewer 3: What are your thoughs on adding an "integration testing" section here (or as a recipe)?
This would test the component consuming the signal store.

Similar to https://ngrx.io/guide/store/testing

Author: Yes of course. Will do that.

A Signal Store is a simple Angular Service, so the same testing techniques you use for services apply to Signal Stores as well. This guide focuses on providing examples for common testing scenarios.

A challenging part of testing is knowing how to handle asynchronous tasks and mocking. The examples use Jest (v29.5), but the same principles apply to other testing frameworks.

There are two main scenarios for testing:

1. Testing the `signalStore` itself,
2. Testing a component or service that uses the Signal Store.

> In the first scenario, you’ll mock the dependencies of the Signal Store, while in the second scenario, you’ll mock the Signal Store itself.

Reviewer 4: There is a third scenario where we test the service by:
1. creating a testing component
2. injecting the service
3. mocking only the service dependencies (not the signal store)

In this way, we can make assertions about what user behavior the service provides without testing the implementation details of the service.

Scenario one above tests pure implementation details and tells us very little about how the service will be used alongside user behavior. Scenario two above tests user behavior, but depending on how broad the scope of the service is may not result in 100% coverage of the branches/functions/etc of the service in that single component.

I think there's a good opportunity to present an opinion about how services can be tested in general (via this third scenario).

I think Tim is hinting at this in his [comment about integration testing](https://github.com/ngrx/platform/pull/4461/files#r1704466865) but I don't want to put words in his mouth

Author: Yes, I thought it is not necessary, but you two have convinced me. I will add it.

```typescript
export const MoviesStore = signalStore(
  withState({
    movies: [
      // ... entries
    ],
  })
);
```

The required addition is that the internal `TestingModule` must provide the `MoviesStore`.

_movies.store.spec.ts_

```typescript
import { MoviesStore } from './movies.store';

describe('MoviesStore', () => {
  it('should verify that three movies are available', () => {
    TestBed.configureTestingModule({
      providers: [MoviesStore],
    });

    const store = TestBed.inject(MoviesStore);

    expect(store.movies()).toHaveLength(3);
  });
});
```

### `withComputed`

Testing derived values of `withComputed` is also straightforward.

_movies.store.ts_

```typescript
export const MoviesStore = signalStore(
  withState({
    movies: [
      // ... entries
    ],
  }),
  withComputed((state) => ({
    moviesCount: computed(() => state.movies().length),
  }))
);
```

_movies.store.spec.ts_

```typescript
import { MoviesStore } from './movies.store';

describe('MoviesStore', () => {
  it('should verify that three movies are available', () => {
    const store = TestBed.inject(MoviesStore);

    expect(store.moviesCount()).toBe(3);
  });
});
```

> ### `withMethods`, DI, and asynchronous tasks

Reviewer 3: I think this is a great example to add, but I think the learning curve is steep from going from "computed" to "async methods".
What about introducing an extra step that shows how to test "simpler" methods that update state in a synchronous way?
The main reason is that I'm afraid that devs will be scared away with the jest fake timers 😅

    loading: false,
  }),
  withMethods((store) => {
    const moviesService = store.inject(MoviesService);

    return {
      async load(studio: string) {
        this.patchState({ loading: true });
        const movies = await moviesService.loadMovies(studio);
        this.patchState(store, { studio, movies, loading: false });
      },
    };
  })
);
```

We will mock the `MoviesService` in our test and let the implementation return the result as a `Promise`.

```typescript
describe('MoviesStore', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(async () => {
    await jest.runOnlyPendingTimersAsync();
    jest.useRealTimers();
  });

  it('should load movies of Warner Bros', async () => {
    const moviesService = {
      load: () =>
        Promise.resolve([
          { id: 1, name: 'Harry Potter' },
          { id: 2, name: 'The Dark Knight' },
        ]),
    };

    TestBed.configureTestingModule({
      providers: [
        {
          provide: MoviesService,
          useValue: moviesService,
        },
      ],
    });

    const store = TestBed.inject(MoviesStore);
    store.load('Warner Bros');
    expect(store.loading()).toBe(true);

>     await jest.runOnlyPendingTimersAsync();

Author: Yup, I would agree if those functions were also available in zoneless applications. Parts of my lending towards the jest-specific API are due to some internal information. Maybe we can discuss this privately or in the GDE channel?

    expect(store.moviesCount()).toBe(2);
    expect(store.loading()).toBe(false);
  });
});
```

<div class="alert is-helpful">

**Note:**: You don't have to manually mock your dependencies. You can choose among libraries like ng-mocks, @testing-library/angular, [jest|jasmine]-auto-spies, etc.

</div>

> We use Jest's tools to manage asynchronous tasks. We avoid using `fakeAsync` or `waitForAsync` as they depend on zone.js and are incompatible with zoneless applications.

Author: Sorry, @markostanimirovic, I just spotted your comment here and answered the above.

I think we should get a clarification from the official side first and then decide if we stick to `fakeAsync` or go with the native ones.

  });
});
```

Be aware of the glitch-free effect when using Signals. The `rxMethod` relies on `effect`, which might need to be triggered manually via `TestBed.flushEffects()`.

If the mocked `MovieService` operates synchronously, the following test would fail without calling `TestBed.flushEffects()`.

```typescript
describe('MoviesStore', () => {
  // ... beforeEach, and afterEach omitted

  it('should depend on flushEffects because of synchronous execution', () => {
    const moviesService = {
      load: jest.fn((studio: string) => of([studio === 'Warner Bros' ? { id: 1, name: 'Harry Potter' } : { id: 2, name: 'Jurassic Park' }])),
    };

    TestBed.configureTestingModule({
      providers: [
        {
          provide: MoviesService,
          useValue: moviesService,
        },
      ],
    });

    const store = TestBed.inject(MoviesStore);
    const studio = signal('Warner Bros');
    store.load(studio);
    TestBed.flushEffects(); // required
    expect(store.movies()).toEqual([{ id: 1, name: 'Harry Potter' }]);

    studio.set('Universal');
    TestBed.flushEffects(); // required
    expect(store.movies()).toEqual([{ id: 2, name: 'Jurassic Park' }]);
  });
});
```

> ### Custom extensions

Reviewer 3: A thought: would this be better if it would be a recipe?
This page is lengthy (which isn't a bad thing), and I think the main focus should be on testing a signal store.

Reviewer: We can also add the "Testing" section to the "Custom Store Features" page, but I'd keep it here to have the full testing guide of SignalStore and related APIs in a single place.

Author: We don't have or require special Testing APIs, that need documentation. The SignalStore is an Angular service and service testing is covered by the official Angular docs.
As Tim said, it is, therefore, more a recipe page than a testing guide. 

I see three main sections:
1. Testing the SignalStore
2. Mocking the SignalStore
3. Integration Tests

I would see "Testing Custom Store Features" as a subchapter of "Testing the SignalStore".

  const moviesStore = {
    movies: signal(new Array<Movie>()),
    loading: signal(false),
    load,
  };

  TestBed.configureTestingModule({
    imports: [MoviesComponent],
    providers: [
      {
        provide: MoviesStore,
        useValue: moviesStore,
      },
    ],
  });

  const fixture = TestBed.createComponent(MoviesComponent);
  fixture.autoDetectChanges(true);

  const studio = load.mock.calls[0][0];
  const input: HTMLInputElement = fixture.debugElement.query(By.css('input')).nativeElement;

  expect(studio()).toBe('');

  input.value = 'Warner Bros';
  input.dispatchEvent(new Event('input'));
  expect(studio()).toBe('Warner Bros');

  moviesStore.movies.set([
    { id: 1, name: 'Harry Potter' },
    { id: 2, name: 'The Dark Knight' },
  ]);
  fixture.detectChanges();

  const movieNames = fixture.debugElement.queryAll(By.css('p')).map((el) => el.nativeElement.textContent);
  expect(movieNames).toEqual(['1: Harry Potter', '2: The Dark Knight']);
});
```

The test mocks only the properties and methods that are used by the component in that particular test. Even if a Signal Store has additional methods, it is not necessary to mock all of them.

> ### ng-mocks

Reviewer 3: I make the assumption that this is added because of the popularity of ng-mocks.
But, I don't think this adds much value here and can be removed.
We already provide good examples and have a mention/link to `ng-mocks`.

```typescript
it('should show movies (spy)', () => {
  TestBed.configureTestingModule({
    imports: [MoviesComponent],
    providers: [
      {
        provide: MoviesService,
        useValue: {},
      },
    ],
  });

  const moviesStore = TestBed.inject(MoviesStore);
  const loadSpy = jest.spyOn(moviesStore, 'load');
  const fixture = TestBed.createComponent(MoviesComponent);

  fixture.autoDetectChanges(true);

  const studio = loadSpy.mock.calls[0][0];
  if (studio instanceof Observable || typeof studio === 'string') {
    throw new Error('Expected signal');
  }

  const input: HTMLInputElement = fixture.debugElement.query(
    By.css('input')
  ).nativeElement;

  expect(studio()).toBe('');

  input.value = 'Warner Bros';
  input.dispatchEvent(new Event('input'));
  expect(studio()).toBe('Warner Bros');

  patchState(moviesStore, {
      movies:
        [
          {id: 1, name: 'Harry Potter'},
          {id: 2, name: 'The Dark Knight'},
        ]
    }
  );

  fixture.detectChanges();

  const movies = fixture.debugElement.queryAll(By.css('p')).map((el) => el.nativeElement.textContent);
  expect(movies).toEqual(['1: Harry Potter', '2: The Dark Knight']);
})
```

> This version requires that the test can modify the state, even if it is protected.

Reviewer: ```suggestion
This version requires the `MoviesStore` state to be unprotected.
```

> We are currently evaluating options for test helpers that allow state modification for protected states without introducing new mocking features that are already covered by existing testing frameworks.

Reviewer: This sentence can be removed from the official guide. Opening an issue on this topic would be a better option in my opinion.

> Additionally, there is a community project by Gergely Szerovay that provides a comprehensive mocking library for Signal Stores. You can find more details here: https://www.angularaddicts.com/p/how-to-mock-ngrx-signal-stores

Reviewer: Same here. To keep this guide concise, we can add the "Community Plugins" page and share community work there.
